File: src/main/java/one/pkg/modpublish/api/ModrinthAPI.java
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package one.pkg.modpublish.api;

import com.google.gson.JsonObject;
import com.intellij.openapi.project.Project;
import okhttp3.*;
import one.pkg.modpublish.data.internel.ModInfo;
import one.pkg.modpublish.data.internel.PublishData;
import one.pkg.modpublish.data.internel.RequestStatus;
import one.pkg.modpublish.data.local.DependencyInfo;
import one.pkg.modpublish.data.local.LauncherInfo;
import one.pkg.modpublish.data.local.MinecraftVersion;
import one.pkg.modpublish.data.network.modrinth.ModrinthData;
import one.pkg.modpublish.data.network.modrinth.ProjectRelation;
import one.pkg.modpublish.data.result.PublishResult;
import one.pkg.modpublish.settings.properties.PID;
import one.pkg.modpublish.util.io.JsonParser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;

public class ModrinthAPI extends API {
    private static final String URL = "https://api.modrinth.com/v2/";

    @Override
    public @NotNull String getID() {
        return "Modrinth";
    }

    @Override
    public void updateABServer() {

    }

    @Override
    public boolean getABServer() {
        return true;
    }

    @Override
    @SuppressWarnings("all")
    public @NotNull PublishResult createVersion(@NotNull PublishData data, @NotNull Project project) {
        Request.Builder requestBuilder = getFormRequest(getRequestBuilder("version", project));

        MultipartBody.Builder bodyBuilder = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("data", createJsonBody(data, project));

        for (int i = 0; i < data.files().length; i++) {
            File f = data.files()[i];
            String key = i == 0 ? f.getName() + "-primary" : f.getName() + "-" + (i - 1);
            bodyBuilder = bodyBuilder.addFormDataPart(key, f.getName(), RequestBody.create(f, MediaType.get("application/java-archive")));
        }

        Request request = requestBuilder.post(bodyBuilder.build()).build();

        try (Response resp = NetworkUtil.client.newCall(request).execute()) {
            @Nullable String status = getStatus(resp);
            return status != null ? PublishResult.create(this, status) : PublishResult.empty();
        } catch (IOException e) {
            return PublishResult.create(this, e.getMessage());
        }
    }

    @Override
    @NotNull
    String createJsonBody(@NotNull PublishData data, @NotNull Project project) {
        var json = ModrinthData.builder().releaseChannel(data.releaseChannel())
                .projectId(PID.ModrinthModID.get(project))
                .versionBody(data.changelog())
                .status(RequestStatus.Listed)
                .featured(true)
                //.primaryFile(data.file())
                .name(data.versionName())
                .versionNumber(data.versionNumber());
        for (File file : data.files()) json.filePart(file);
        for (LauncherInfo l : data.loaders()) json.loader(l);
        for (MinecraftVersion version : data.minecraftVersions()) json.gameVersion(version);
        for (DependencyInfo d : data.dependencies()) {
            ModInfo info = d.getModrinthModInfo();
            if (info == null || info.modid() == null || info.modid().isBlank() ||
                    info.slug() == null || info.slug().isBlank()) continue;
            json.dependency(
                    ProjectRelation.create(info.modid(), d.getType())
            );
        }
        return json.build().toJson();
    }

    @Override
    public @NotNull ModInfo getModInfo(@NotNull String modid, @NotNull Project project) {
        Request req = getJsonRequest(getRequestBuilder("project/" + modid, project)).get().build();
        try (Response resp = NetworkUtil.client.newCall(req).execute()) {
            @Nullable String status = getStatus(resp);
            if (status != null) return ModInfo.of(status);
            JsonObject object = JsonParser.getJsonObject(resp.body().byteStream());
            return ModInfo.of(modid, object.get("title").getAsString(), object.get("slug").getAsString());
        } catch (IOException e) {
            return ModInfo.of(e.getMessage());
        }
    }

    @NotNull
    Request.Builder getRequestBuilder(@NotNull String url, @NotNull Project project) {
        return getBaseRequestBuilder()
                .header("Authorization", PID.ModrinthToken.getProtect(project).data())
                .url(URL + url);
    }
}

File: src/main/kotlin/one/pkg/modpublish/api/ModrinthAPI.kt
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package one.pkg.modpublish.api

import com.intellij.openapi.project.Project
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import one.pkg.modpublish.api.NetworkUtil.client
import one.pkg.modpublish.data.internal.ModInfo
import one.pkg.modpublish.data.internal.PublishData
import one.pkg.modpublish.data.internal.RequestStatus
import one.pkg.modpublish.data.network.modrinth.ModrinthData
import one.pkg.modpublish.data.network.modrinth.ProjectRelation
import one.pkg.modpublish.data.result.PublishResult
import one.pkg.modpublish.settings.properties.PID
import one.pkg.modpublish.util.io.JsonParser.getJsonObject
import one.pkg.modpublish.util.io.JsonParser.toJson
import java.io.IOException

class ModrinthAPI : API() {
    override val id: String get() = "Modrinth"

    private var aBServer = false

    override fun getAB(): Boolean = aBServer
    override fun updateAB() {}

    override fun createVersion(data: PublishData, project: Project): PublishResult {
        val requestBuilder = getFormRequest(getRequestBuilder("version", project))
        val body = MultipartBody.Builder().setType(MultipartBody.FORM).apply {
            addFormDataPart("data", createJsonBody(data, project))
            data.files.forEachIndexed { i, file ->
                val key = if (i == 0) "${file.name}-primary" else "${file.name}-${i - 1}"
                addFormDataPart(key, file.name, file.asRequestBody("application/java-archive".toMediaType()))
            }
        }.build()

        val request = requestBuilder.post(body).build()

        return try {
            client.newCall(request).execute().use { resp ->
                getStatus(resp)?.let { return PublishResult.create(this, it) }
                PublishResult.EMPTY
            }
        } catch (e: IOException) {
            PublishResult.create(this, e.message)
        }
    }

    override fun createJsonBody(data: PublishData, project: Project): String {
        return ModrinthData().apply {
            this.releaseChannel(data.releaseChannel)
            this.projectId = PID.ModrinthModID.get(project)
            this.versionBody = data.changelog
            this.status(RequestStatus.Listed)
            this.featured = true
            this.name = data.versionName
            this.versionNumber = data.versionNumber
            data.files.forEach { this.filePart(it) }
            data.loaders.forEach { this.loader(it) }
            data.minecraftVersions.forEach { this.gameVersion(it) }
            data.dependencies.forEach { d ->
                val info = d.modrinthModInfo
                if (!info?.modid.isNullOrBlank() && !info.slug.isNullOrBlank()) {
                    this.dependency(ProjectRelation.create(info.modid, d.type))
                }
            }
        }.toJson()
    }

    override fun getModInfo(modid: String, project: Project): ModInfo {
        val request = getJsonRequest(getRequestBuilder("project/$modid", project)).get().build()
        return try {
            client.newCall(request).execute().use { resp ->
                getStatus(resp)?.let { return ModInfo.of(it) }
                val obj = resp.body.byteStream().getJsonObject()
                ModInfo.of(modid, obj.get("title").asString, obj.get("slug").asString)
            }
        } catch (e: IOException) {
            ModInfo.of(e.message)
        }
    }

    fun getRequestBuilder(url: String, project: Project): Request.Builder =
        baseRequestBuilder.header("Authorization", PID.ModrinthToken.getProtect(project).data)
            .url(URL + url)

    companion object {
        private const val URL = "https://api.modrinth.com/v2/"
    }
}